File: learning-factory/warehouse/artifacts/projects/languages/sprig-bytecode-vm/job_project_bytecode_vertical_v1/attempt-002/debugging/parser-associativity/buggy/tinyvm/compiler.py
from __future__ import annotations

from dataclasses import dataclass

from .model import Assign, Binary, Block, CompileError, Expr, If, Let, Literal, Print, Program, Unary, Variable, While


@dataclass(frozen=True)
class Instruction:
    opcode: str
    operand: int | str | None = None


@dataclass(frozen=True)
class BytecodeProgram:
    instructions: tuple[Instruction, ...]


class Compiler:
    def __init__(self) -> None:
        self.code: list[Instruction] = []

    def compile(self, program: Program) -> BytecodeProgram:
        for statement in program.statements:
            self._statement(statement)
        self._emit("HALT")
        return BytecodeProgram(tuple(self.code))

    def _emit(self, opcode: str, operand: int | str | None = None) -> int:
        self.code.append(Instruction(opcode, operand))
        return len(self.code) - 1

    def _patch(self, location: int, target: int) -> None:
        instruction = self.code[location]
        if instruction.opcode not in {"JUMP", "JUMP_IF_FALSE", "JUMP_IF_TRUE"}:
            raise CompileError("attempted to patch a non-jump instruction")
        self.code[location] = Instruction(instruction.opcode, target)

    def _statement(self, statement) -> None:
        self._emit("TICK")
        if isinstance(statement, Let):
            self._expression(statement.initializer)
            self._emit("DEFINE", statement.name)
            return
        if isinstance(statement, Assign):
            self._expression(statement.value)
            self._emit("STORE", statement.name)
            return
        if isinstance(statement, Print):
            self._expression(statement.expression)
            self._emit("PRINT")
            return
        if isinstance(statement, Block):
            for child in statement.statements:
                self._statement(child)
            return
        if isinstance(statement, If):
            self._expression(statement.condition)
            false_jump = self._emit("JUMP_IF_FALSE", -1)
            self._statement(statement.then_branch)
            end_jump = self._emit("JUMP", -1)
            self._patch(false_jump, len(self.code))
            if statement.else_branch is not None:
                self._statement(statement.else_branch)
            self._patch(end_jump, len(self.code))
            return
        if isinstance(statement, While):
            loop_start = len(self.code)
            self._expression(statement.condition)
            exit_jump = self._emit("JUMP_IF_FALSE", -1)
            self._statement(statement.body)
            self._emit("JUMP", loop_start)
            self._patch(exit_jump, len(self.code))
            return
        raise CompileError(f"unsupported statement: {type(statement).__name__}")

    def _expression(self, expression: Expr) -> None:
        self._emit("TICK")
        if isinstance(expression, Literal):
            self._emit("CONST", expression.value)
            return
        if isinstance(expression, Variable):
            self._emit("LOAD", expression.name)
            return
        if isinstance(expression, Unary):
            if expression.operator == "-" and isinstance(expression.right, Literal) and expression.right.value == 2 ** 63:
                self._emit("TICK")
                self._emit("CONST", -(2 ** 63))
                return
            self._expression(expression.right)
            self._emit({"-": "NEG", "!": "NOT"}[expression.operator])
            return
        if isinstance(expression, Binary) and expression.operator == "&&":
            self._expression(expression.left)
            false_jump = self._emit("JUMP_IF_FALSE", -1)
            self._expression(expression.right)
            self._emit("BOOL")
            end_jump = self._emit("JUMP", -1)
            self._patch(false_jump, len(self.code))
            self._emit("CONST", 0)
            self._patch(end_jump, len(self.code))
            return
        if isinstance(expression, Binary) and expression.operator == "||":
            self._expression(expression.left)
            true_jump = self._emit("JUMP_IF_TRUE", -1)
            self._expression(expression.right)
            self._emit("BOOL")
            end_jump = self._emit("JUMP", -1)
            self._patch(true_jump, len(self.code))
            self._emit("CONST", 1)
            self._patch(end_jump, len(self.code))
            return
        if isinstance(expression, Binary):
            self._expression(expression.left)
            self._expression(expression.right)
            opcode = {
                "+": "ADD", "-": "SUB", "*": "MUL", "/": "DIV", "%": "MOD",
                "==": "EQ", "!=": "NE", "<": "LT", "<=": "LE", ">": "GT", ">=": "GE",
            }.get(expression.operator)
            if opcode is None:
                raise CompileError(f"unsupported operator: {expression.operator}")
            self._emit(opcode)
            return
        raise CompileError(f"unsupported expression: {type(expression).__name__}")


def compile_program(program: Program) -> BytecodeProgram:
    return Compiler().compile(program)

File: learning-factory/warehouse/artifacts/projects/build-your-own-x/programming-language/a-compiler-from-scratch-84427e53fe/job_byox_build_v1_f3e0d99cf97b485202190e3be4a0e8f6/attempt-001/sealed/reference/lib/pebble/compiler.rb
# frozen_string_literal: true

module Pebble
  class Compiler
    BINARY_OPCODES = {
      PLUS: :ADD,
      MINUS: :SUB,
      STAR: :MUL,
      SLASH: :DIV,
      PERCENT: :MOD,
      EQUAL_EQUAL: :EQ,
      BANG_EQUAL: :NE,
      LESS: :LT,
      LESS_EQUAL: :LE,
      GREATER: :GT,
      GREATER_EQUAL: :GE
    }.freeze

    UNARY_OPCODES = {
      MINUS: :NEG,
      BANG: :NOT
    }.freeze

    def compile(ast)
      @instructions = []
      @scopes = [{}]
      @next_slot = 0

      unless ast.is_a?(Hash) && ast[:type] == :program && ast[:statements].is_a?(Array)
        compile_error("expected a program AST", ast.is_a?(Hash) ? ast[:token] : nil)
      end

      ast[:statements].each { |node| compile_statement(node) }
      emit(:HALT)
      Program.new(@instructions, @next_slot)
    end

    private

    def compile_statement(node)
      validate_node(node)
      case node[:type]
      when :let
        compile_let(node)
      when :assign
        compile_expression(node[:value])
        emit(:STORE, resolve(node[:name], node[:token]))
      when :print
        compile_expression(node[:value])
        emit(:PRINT)
      when :if
        compile_if(node)
      when :while
        compile_while(node)
      else
        compile_error("unknown statement node #{node[:type].inspect}", node[:token])
      end
    end

    def compile_let(node)
      name = validate_name(node[:name], node[:token])
      if @scopes.last.key?(name)
        compile_error("variable '#{name}' is already declared in this scope", node[:token])
      end

      compile_expression(node[:value])
      slot = @next_slot
      @next_slot += 1
      @scopes.last[name] = slot
      emit(:STORE, slot)
    end

    def compile_if(node)
      compile_expression(node[:condition])
      false_jump = emit(:JUMP_IF_FALSE, nil)
      compile_block(node[:then_body], node[:token])

      else_body = node[:else_body]
      unless else_body.is_a?(Array)
        compile_error("if else_body must be an Array", node[:token])
      end
      if else_body.empty?
        patch_jump(false_jump, @instructions.length)
      else
        end_jump = emit(:JUMP, nil)
        patch_jump(false_jump, @instructions.length)
        compile_block(else_body, node[:token])
        patch_jump(end_jump, @instructions.length)
      end
    end

    def compile_while(node)
      loop_start = @instructions.length
      compile_expression(node[:condition])
      exit_jump = emit(:JUMP_IF_FALSE, nil)
      compile_block(node[:body], node[:token])
      emit(:JUMP, loop_start)
      patch_jump(exit_jump, @instructions.length)
    end

    def compile_block(statements, token)
      unless statements.is_a?(Array)
        compile_error("block body must be an Array", token)
      end

      @scopes << {}
      begin
        statements.each { |statement| compile_statement(statement) }
      ensure
        @scopes.pop
      end
    end

    def compile_expression(node)
      validate_node(node)
      case node[:type]
      when :literal
        value = node[:value]
        unless value.is_a?(Integer) || value.equal?(true) || value.equal?(false)
          compile_error("invalid literal", node[:token])
        end
        emit(:CONST, value)
      when :variable
        emit(:LOAD, resolve(node[:name], node[:token]))
      when :unary
        opcode = UNARY_OPCODES[node[:operator]]
        compile_error("unknown unary operator", node[:token]) unless opcode
        compile_expression(node[:operand])
        emit(opcode)
      when :binary
        opcode = BINARY_OPCODES[node[:operator]]
        compile_error("unknown binary operator", node[:token]) unless opcode
        compile_expression(node[:left])
        compile_expression(node[:right])
        emit(opcode)
      else
        compile_error("unknown expression node #{node[:type].inspect}", node[:token])
      end
    end

    def resolve(name, token)
      name = validate_name(name, token)
      (@scopes.length - 1).downto(0) do |index|
        return @scopes[index][name] if @scopes[index].key?(name)
      end
      compile_error("variable '#{name}' is not declared", token)
    end

    def validate_name(name, token)
      return name if name.is_a?(String) && name.match?(/\A[A-Za-z_][A-Za-z0-9_]*\z/)

      compile_error("invalid variable name", token)
    end

    def validate_node(node)
      return if node.is_a?(Hash)

      compile_error("expected an AST node", nil)
    end

    def emit(opcode, *operands)
      index = @instructions.length
      @instructions << [opcode, *operands]
      index
    end

    def patch_jump(index, target)
      @instructions[index][1] = target
    end

    def compile_error(message, token)
      if token && token.respond_to?(:line) && token.respond_to?(:column)
        raise CompileError.new(message, token.line, token.column)
      end
      raise CompileError.new(message, 1, 1)
    end
  end
end
